### PR TITLE
stability: ensure AdminSplit and ChangeReplicas begin txn first

### LINKED ISSUE #9287: storage: Range addressing updated too soon
The first batches of the [ChangeReplicas](https://github.com/cockroachdb/cockroach/blob/f48730d4d437c5b901dc8e8cd7b37bb9155fe848/storage/replica_command.go#L3086-L3097) and [AdminSplit](https://github.com/cockroachdb/cockroach/blob/f48730d4d437c5b901dc8e8cd7b37bb9155fe848/storage/replica_command.go#L2203-L2216) transactions do too much. They update both range descriptors and the meta records. By contrast, [AdminMerge](https://github.com/cockroachdb/cockroach/blob/f48730d4d437c5b901dc8e8cd7b37bb9155fe848/storage/replica_command.go#L2660) makes a point of running a batch to modify the left range descriptor before it does anything else. 

This matters because of the way batches are split in the KV layer: Sub-batches are executed in order of their keys, not their position in the batch. This means that the updates to the meta records run first, creating intents before the transaction record is created on the local range descriptor. These intents are unusually vulnerable: until a transaction record has been written, the transaction [loses all pushes](https://github.com/cockroachdb/cockroach/blob/f48730d4d437c5b901dc8e8cd7b37bb9155fe848/storage/replica_command.go#L1285), even those for inconsistent reads (as used for `RangeLookup`. These turn into `PUSH_TOUCH` pushes). (This is not a well-exercised code path; [comments](https://github.com/cockroachdb/cockroach/blob/f48730d4d437c5b901dc8e8cd7b37bb9155fe848/storage/intent_resolver.go#L287-L296) indicate that the expected way that `PUSH_TOUCH` could abort is when the transaction has been abandoned).

AdminSplit and ChangeReplicas transactions may be getting aborted more often than they should be because of this. There is a simple solution: refactor the batch (as in AdminMerge) so the first batch *only* includes the range descriptor where the transaction record is to live. Then the transaction record will be written and inconsistent reads will no longer interfere with the transaction in progress.

Meta records are also special in that uncommitted intents are (sometimes) used. Updating these records as late in the transaction as possible will minimize the exposure of these intents while they are in the uncommitted state. (This is not a correctness issue, but the longer the commits live before the transaction is either committed or aborted, the more likely we are to have unnecessary RPC retries)

I discovered this while investigating #9265. We should probably refrain from fixing this until that issue is more fully understood; it is possible that this behavior is related to the cause of that bug.

### PR DESCRIPTION
**Pull request by spencerkimball**
Reorganizes the transactions done in `AdminSplit` and `ChangeReplicas`
to look more like `AdminMerge`, where we insist on an initial batch
to guarantee that `BeginTransaction` is run before creating any
intents on other ranges.

Also now set `OrigTimestamp` in the transaction record which is
written in the event that no transaction record exists on a push.
This missing value was allowing the gc queue to garbage collect
an aborted transaction without waiting the normal 1 hour grace period.

Fixes #9287

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9374)

### CODE REVIEW

File: storage/client_split_test.go
Comment (tamird) [inline storage/client_split_test.go]: this should mention that an aborted record is written.

File: storage/client_split_test.go
Comment (tamird) [inline storage/client_split_test.go]: broken sentence

File: storage/client_split_test.go
Comment (tamird) [inline storage/client_split_test.go]: i think you want to put this on the context before creating the test store.

File: storage/client_split_test.go
Comment (tamird) [inline storage/client_split_test.go]: `timeutil.Since`

File: storage/client_split_test.go
Comment (tamird) [inline storage/client_split_test.go]: can use `Key.Equal`

File: storage/client_split_test.go
Comment (tamird) [inline storage/client_split_test.go]: why does that matter?

File: storage/client_split_test.go
Comment (tamird) [inline storage/client_split_test.go]: should these be returns? i don't understand which part of this is expected to always succeed vs which part needs the retry loop.

File: storage/helpers_test.go
Comment (tamird) [inline storage/helpers_test.go]: there's another place that returns such an error, and that one does:
```
	cfg, ok := s.Gossip().GetSystemConfig()
	if !ok {
		return errors.Errorf("%s: system config not yet available", s)
	}
```

File: storage/helpers_test.go
Comment (tamird) [inline storage/helpers_test.go]: s.Ctx()?

File: storage/replica_command.go
Comment (tamird) [inline storage/replica_command.go]: I believe this should be `txn.NewBatch()` here and below.

File: storage/replica_command.go
Comment (tamird) [inline storage/replica_command.go]: ditto

**Review by tamird — COMMENTED** [review]
The sleeps in that test look pretty sketchy to me, but I assume you've discussed this with @tschottdorf.

Please be sure to stress this before merging.

File: storage/client_split_test.go
Comment (spencerkimball) [inline storage/client_split_test.go]: Done

File: storage/client_split_test.go
Comment (spencerkimball) [inline storage/client_split_test.go]: Done

File: storage/client_split_test.go
Comment (spencerkimball) [inline storage/client_split_test.go]: I want that first split to avoid the testing command filter.

File: storage/client_split_test.go
Comment (spencerkimball) [inline storage/client_split_test.go]: Done

File: storage/client_split_test.go
Comment (spencerkimball) [inline storage/client_split_test.go]: Done

File: storage/client_split_test.go
Comment (spencerkimball) [inline storage/client_split_test.go]: If the GC doesn't finish, then the test won't fail as expected before these changes went in.

File: storage/client_split_test.go
Comment (spencerkimball) [inline storage/client_split_test.go]: These are both test errors. The succeeds construct only runs as long as there are any errors getting the meta2 record due to intents.

File: storage/helpers_test.go
Comment (spencerkimball) [inline storage/helpers_test.go]: Done

File: storage/helpers_test.go
Comment (spencerkimball) [inline storage/helpers_test.go]: Done

File: storage/replica_command.go
Comment (spencerkimball) [inline storage/replica_command.go]: Changed.

File: storage/replica_command.go
Comment (spencerkimball) [inline storage/replica_command.go]: Done

**Comment by spencerkimball** [comment]
+cc @cockroachdb/stability

File: storage/client_split_test.go
Comment (tamird) [inline storage/client_split_test.go]: Oh, you have a log.Errorf there. I thought I saw `t.Errorf` before.

**Review by tamird — APPROVED** [review]
Well, Github code review still sucks.

**Comment by tschottdorf** [comment]
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/client_split_test.go, line 1585 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRe_Ot50hei5cOmmzhQ:-KRe_Ot50hei5cOmmzhR:b-6ch1qj) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1585)):*
> ```Go
> // updates to meta addressing records to proceed. Because the test
> // performs an initial split at the SystemPrefix, the dist sender ends
> // up issuing a range lookup request after failing to send the entire
> ```

Just make the read of the intent explicit, which also allows you to avoid the sleep - you wait for the ResolveIntent which aborts the intent on the meta2 key (triggered by your Get).

---

*[storage/client_split_test.go, line 1626 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReZx10pIHYBF1Og9tQ:-KReZx10pIHYBF1Og9tR:b8ju4gl) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1626)):*
> ```Go
> 			filterArgs.Req.Header().Key.Equal(keys.RangeMetaKey(splitKey)) {
> 			select {
> 			case wroteMeta2 <- struct{}{}:
> ```

Looks like a potential deadlock here if the txn has to retry.

---

*[storage/client_split_test.go, line 1655 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReZ_3pG_K6og-QubGu:-KReZ_3pG_K6og-QubGv:b-xeuelu) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1655)):*
> ```Go
> 	if err := store.ManualReplicaGC(store.LookupReplica(splitKey, nil)); err != nil {
> 		// TODO(spencer): test fatal here after the gc queue correctly
> 		// uses dist sender, because there is a chance the gc spans ranges.
> ```

s/dist sender/DistSender/g

What is the issue here? You could always `SucceedsWithin`.

---

*[storage/replica_command.go, line 1311 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReX7kDW8myOxP_teao:-KReX7kDW8myOxP_teap:b-21zor0) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/replica_command.go#L1311)):*
> ```Go
> 		// to receive dummy values.
> 		reply.PusheeTxn.TxnMeta = args.PusheeTxn
> 		reply.PusheeTxn.OrigTimestamp = args.Now
> ```

I know we discussed before that you would put this here, but would you mind removing it if it's not necessary for your test to pass? I also had to fix this in my WIP to make the test happy (i.e. I exercise this unit-test style) and we can avoid a merge conflict that way.

---

*[storage/replica_command.go, line 2295 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReXQyMkN0EAYZ1CQfX:-KReXQyNtVyNb4nuTKS1:b-azxe7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/replica_command.go#L2295)):*
> ```Go
> 			// This prevents a race where GC encounters intents without a txn
> 			// record and deletes them, and then allows the begin txn to
> 			// recreate the txn and continue (see #9265).
> ```

Good to mention the race, but really (assuming other bugs fixed) we do this here because it's economical to not provoke our splits being aborted early - I would put more emphasis on that or it could seem that this is a workaround, where it's really a performance fix (and secondarily a temporary workaround).

---

*[storage/replica_command.go, line 2306 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReXw7VmUl3YjDBa6C-:-KReXw7WSKeqCfA3L1ei:b27s2wu) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/replica_command.go#L2306)):*
> ```Go
> 
> 		// Log the split into the range event log.
> 		// TODO(spencer): event logging API should accept a vanilla DB
> ```

This TODO doesn't seem to check out - this needs to be done with the txn (because retries), so unless you're suggesting using the txn in parallel (which is surely going to explode) there's nothing to do here. However, perhaps `logSplit` can feed into a batch so that we don't have to run an extra call for it (and it would automagically benefit from future DistSender parallelism).

---

*[storage/replica_command.go, line 2342 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReYq82MHSfVmfBOg4T:-KReYq83XlnPy56yFSSF:b-2a4pw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/replica_command.go#L2342)):*
> ```Go
> 		if err := txn.Run(b); err != nil {
> 			if _, ok := err.(*roachpb.ConditionFailedError); ok {
> 				return errors.Errorf("conflict updating range descriptors")
> ```

disambiguate which step this was in (compared to the above one for the left desc).

---

*[storage/replica_command.go, line 3200 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReZ2xKr773bKjidesN:-KReZ2xKr773bKjidesO:bpcf0kl) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/replica_command.go#L3200)):*
> // Run transaction up to this point to create the transaction record early (see #9265).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374)*

**Comment by spencerkimball** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/client_split_test.go, line 1585 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRe_Ot50hei5cOmmzhQ:-KReinHADiK7lce44Trh:b-aynuv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1585)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Just make the read of the intent explicit, which also allows you to avoid the sleep - you wait for the ResolveIntent which aborts the intent on the meta2 key (triggered by your Get).
</details>
It just ends up sending another RangeLookup through when one has already been sent and it doesn't help with the sleep because unfortunately the sleep is there to allow the write of the meta2 records to go from the testing command filter down through to batch commit on the replicas (this is all necessary for the case of "before the fix", and it's the reason this test only succeeds for the old code 199 out of 200 times on my laptop).

---

*[storage/client_split_test.go, line 1626 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReZx10pIHYBF1Og9tQ:-KRejQJ4O4iMegiHmTLo:b-5ycsow) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1626)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Looks like a potential deadlock here if the txn has to retry.
</details>
Why? This select loop allows the channel to fill up without blocking.

---

*[storage/client_split_test.go, line 1655 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReZ_3pG_K6og-QubGu:-KRejvTjsOjFU7oYy_rf:b-ovwaw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1655)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/dist sender/DistSender/g
> 
> What is the issue here? You could always `SucceedsWithin`.
</details>
Good idea.

---

*[storage/replica_command.go, line 1311 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReX7kDW8myOxP_teao:-KRekW674czTxZwqh38y:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/replica_command.go#L1311)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I know we discussed before that you would put this here, but would you mind removing it if it's not necessary for your test to pass? I also had to fix this in my WIP to make the test happy (i.e. I exercise this unit-test style) and we can avoid a merge conflict that way.
</details>
OK

---

*[storage/replica_command.go, line 2306 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReXw7VmUl3YjDBa6C-:-KRekzS2bW0mAt9H4i2h:b-by58mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/replica_command.go#L2306)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This TODO doesn't seem to check out - this needs to be done with the txn (because retries), so unless you're suggesting using the txn in parallel (which is surely going to explode) there's nothing to do here. However, perhaps `logSplit` can feed into a batch so that we don't have to run an extra call for it (and it would automagically benefit from future DistSender parallelism).
</details>
By "vanilla DB", I meant the batch. Aren't they implementing an interface, or am I misremembering?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374)*

**Comment by tschottdorf** [comment]
Looks good so far, just some test tickling left. BTW, would you mind "acknowledging" comments that you address but don't have a response to? That helps me reviewing. I've done it myself this time.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/client_split_test.go, line 1585 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRe_Ot50hei5cOmmzhQ:-KRenfwjBh9givKJYgoD:b-qa4z3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1585)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> It just ends up sending another RangeLookup through when one has already been sent and it doesn't help with the sleep because unfortunately the sleep is there to allow the write of the meta2 records to go from the testing command filter down through to batch commit on the replicas (this is all necessary for the case of "before the fix", and it's the reason this test only succeeds for the old code 199 out of 200 times on my laptop).
</details>
If instead of sleeping and having the meta2 channel you would just repeatedly read the corresponding meta key until you intercept a ResolveIntent which removes that intent, I think you would get what you want 100% of the time without waiting:

- initiate the split async, but blocking the BeginTransaction (as you do now)
- send RangeLookup to that key yourself until you catch a ResolveIntent(abort=true) in the command filter
- you now know that the split has written its LHS intent and it has been aborted
- GC queue
- unblock the Begin

---

*[storage/client_split_test.go, line 1626 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReZx10pIHYBF1Og9tQ:-KRen4ooKDiGprs39uPX:b-ffqi5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1626)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Why? This select loop allows the channel to fill up without blocking.
</details>
Oh, you're right. Didn't even register.

---

*[storage/client_split_test.go, line 1653 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReou8qCPr6TCbZbfnY:-KReou8qCPr6TCbZbfnZ:bl2aasv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f50287d284a3c95b1f87f7563a665e516eab391/storage/client_split_test.go#L1653)):*
> ```Go
> 
> 	// GC the replica containing the range descriptor records.
> 	util.SucceedsSoon(t, func() error {
> ```

Explain the SucceedsSoon. Looks odd without context.

---

*[storage/replica_command.go, line 2306 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReXw7VmUl3YjDBa6C-:-KRenUDXRk3tgOiM0C9z:b-mvtsj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/replica_command.go#L2306)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> By "vanilla DB", I meant the batch. Aren't they implementing an interface, or am I misremembering?
</details>
I don't think they implement an interface.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-:-KRep3NNOIPS_z1JvyzS:bpky65g)*

**Comment by tamird** [comment]
Reviewed 3 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/client_split_test.go, line 1654 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRerSes1-E4n9rz3Lkh:-KRerSes1-E4n9rz3Lki:b-73gnbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f50287d284a3c95b1f87f7563a665e516eab391/storage/client_split_test.go#L1654)):*
> ```Go
> 	// GC the replica containing the range descriptor records.
> 	util.SucceedsSoon(t, func() error {
> 		if err := store.ManualReplicaGC(store.LookupReplica(splitKey, nil)); err != nil {
> ```

`return store....`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374)*

**Comment by spencerkimball** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/client_split_test.go, line 1585 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRe_Ot50hei5cOmmzhQ:-KReuqHHU3SRt-VdMpIV:bd6er5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1585)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> If instead of sleeping and having the meta2 channel you would just repeatedly read the corresponding meta key until you intercept a ResolveIntent which removes that intent, I think you would get what you want 100% of the time without waiting:
> 
> - initiate the split async, but blocking the BeginTransaction (as you do now)
> - send RangeLookup to that key yourself until you catch a ResolveIntent(abort=true) in the command filter
> - you now know that the split has written its LHS intent and it has been aborted
> - GC queue
> - unblock the Begin
</details>
As mentioned offline, I still maintain that the unittest as is does the "right thing", as it must fail for the previous case and succeed for the current code.

---

*[storage/client_split_test.go, line 1626 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReZx10pIHYBF1Og9tQ:-KRev0d78Neczu3mRRTL:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/client_split_test.go#L1626)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Oh, you're right. Didn't even register.
</details>
Done

---

*[storage/client_split_test.go, line 1653 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReou8qCPr6TCbZbfnY:-KRevOT9wELuoOP9Nm_8:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f50287d284a3c95b1f87f7563a665e516eab391/storage/client_split_test.go#L1653)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Explain the SucceedsSoon. Looks odd without context.
</details>
Done

---

*[storage/client_split_test.go, line 1654 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRerSes1-E4n9rz3Lkh:-KRevXLozQOnnWJman5S:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f50287d284a3c95b1f87f7563a665e516eab391/storage/client_split_test.go#L1654)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `return store....`
</details>
Done

---

*[storage/replica_command.go, line 2306 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KReXw7VmUl3YjDBa6C-:-KRevYYtlayfxyv5YieX:b-tp81h3) ([raw file](https://github.com/cockroachdb/cockroach/blob/347b9695a5a6453c02f2db7d0738d43d60b0d827/storage/replica_command.go#L2306)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I don't think they implement an interface.
</details>
Bugger...well I updated the comment to say "batch".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374)*

**Comment by tschottdorf** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-:-KRezSNjulzTcJj5PCSY:bnfp4nl)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/client_split_test.go, line 1579 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRfSjAwuC38yM3FWLYX:-KRfSjAwuC38yM3FWLYY:b-uc90pg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4686ea21cbc5e3ed2938ec2a1f358df3d27688d/storage/client_split_test.go#L1579)):*
> ```Go
> }
> 
> // TestStoreSplitBeginTxnPushMetaIntentRace prevents regression of
> ```

I thought we were programming in Go, not Objective-C.

---

*[storage/replica_command.go, line 2290 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRfT6U7uvJzp1GihM5x:-KRfT6U7uvJzp1GihM5y:bi0a3v2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4686ea21cbc5e3ed2938ec2a1f358df3d27688d/storage/replica_command.go#L2290)):*
> ```Go
> 			// Commit this batch first to ensure that the transaction record
> 			// is created in the right place (our triggers rely on this),
> 			// but also to ensure the transaction record is created _before_
> ```

I think you could expand on why not doing this does not ensure that the transaction record is created first. The naive reader would assume that when a transaction runs the transaction record is written first. The fact that DistSender can muck with that assumption is not at all clear.

---

*[storage/replica_command.go, line 2339 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRfTd3fxPVAjSPqu-2_:-KRfTd3glktbO6u7HmOE:b-wx9qbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4686ea21cbc5e3ed2938ec2a1f358df3d27688d/storage/replica_command.go#L2339)):*
> ```Go
> 
> 		// Commit txn with final batch (RHS desc and meta).
> 		log.Trace(ctx, "commit txn with batch containing RHS descriptor and meta records")
> ```

Above you use `left descriptor`. Here you use `RHS descriptor`. Consistency please.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374)*

**Comment by spencerkimball** [comment]
*[storage/replica_command.go, line 2339 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRfTd3fxPVAjSPqu-2_:-KRhzqI52MJVusBuU85W:b32w0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4686ea21cbc5e3ed2938ec2a1f358df3d27688d/storage/replica_command.go#L2339)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Above you use `left descriptor`. Here you use `RHS descriptor`. Consistency please.
</details>
OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374)*

**Comment by spencerkimball** [comment]
*[storage/replica_command.go, line 2290 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRfT6U7uvJzp1GihM5x:-KRhzhgGr958MUhqAYFy:b-eowr3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4686ea21cbc5e3ed2938ec2a1f358df3d27688d/storage/replica_command.go#L2290)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think you could expand on why not doing this does not ensure that the transaction record is created first. The naive reader would assume that when a transaction runs the transaction record is written first. The fact that DistSender can muck with that assumption is not at all clear.
</details>
Clarified.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374)*

**Comment by spencerkimball** [comment]
*[storage/client_split_test.go, line 1579 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374#-KRfSjAwuC38yM3FWLYX:-KRhyYqLQXuirUNIxXbv:b-s503mt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4686ea21cbc5e3ed2938ec2a1f358df3d27688d/storage/client_split_test.go#L1579)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I thought we were programming in Go, not Objective-C.
</details>
yeah yeah

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9374)*

**Pull request merged**
